prometheus/registry.go

```diff
@@ -11,232 +11,253 @@
-	// for debugging.)
-	ProtoCompactTextTelemetryContentType = `application/vnd.google.protobuf; proto=io.prometheus.client.MetricFamily; encoding=compact-text`
-
-	// Constants for object pools.
-	numBufs           = 4
-	numMetricFamilies = 1000
-	numMetrics        = 10000
-
 	// Capacity for the channel to collect metrics and descriptors.
 	capMetricChan = 1000
 	capDescChan   = 10
+)
 
-	contentTypeHeader     = "Content-Type"
-	contentLengthHeader   = "Content-Length"
-	contentEncodingHeader = "Content-Encoding"
+// DefaultRegistry is the default registry implicitly used by a number of
+// convenience functions. It has a ProcessCollector and a GoCollector
+// pre-registered.
+var DefaultRegistry = NewRegistry()
 
-	acceptEncodingHeader = "Accept-Encoding"
-	acceptHeader         = "Accept"
-)
+func init() {
+	MustRegister(NewProcessCollector(os.Getpid(), ""))
+	MustRegister(NewGoCollector())
+}
 
-// Handler returns the HTTP handler for the global Prometheus registry. It is
-// already instrumented with InstrumentHandler (using "prometheus" as handler
-// name). Usually the handler is used to handle the "/metrics" endpoint.
-//
-// Please note the issues described in the doc comment of InstrumentHandler. You
-// might want to consider using UninstrumentedHandler instead.
-func Handler() http.Handler {
-	return InstrumentHandler("prometheus", defRegistry)
+// NewRegistry creates a new vanilla Registry without any Collectors
+// pre-registered.
+func NewRegistry() Registry {
+	return &registry{
+		collectorsByID:  map[uint64]Collector{},
+		descIDs:         map[uint64]struct{}{},
+		dimHashesByName: map[string]uint64{},
+	}
 }
 
-// UninstrumentedHandler works in the same way as Handler, but the returned HTTP
-// handler is not instrumented. This is useful if no instrumentation is desired
-// (for whatever reason) or if the instrumentation has to happen with a
-// different handler name (or with a different instrumentation approach
-// altogether). See the InstrumentHandler example.
-func UninstrumentedHandler() http.Handler {
-	return defRegistry
+// NewRegistryWithInjectionHook creates a registry with the provided hook to inject
+// MetricFamilies. The hook is a function that is called whenever metrics are
+// collected. The MetricFamily protobufs returned by the hook function are
+// merged with the metrics collected in the usual way.
+//
+// This is a way to directly inject MetricFamily protobufs managed and owned by
+// the caller. The caller has full responsibility. As no registration of the
+// injected metrics has happened, there was no check at registration-time. If
+// the injection results in inconsistent metrics, the Collect call will return
+// an error. Some problems may even go undetected, like invalid label names in
+// the injected protobufs.
+//
+// The hook function must be callable at any time and concurrently.
+func NewRegistryWithInjectionHook(hook func() []*dto.MetricFamily) Registry {
```

Comment:
In this particular case, there is no 2^n problem, as (incidentally) the two special cases here are effectively mutually exclusive (the pedantic checks are useless if you set an injection hook, which has no descriptor and cannot be checked).

Having said that, I thought a while about this. Conclusions:

- Both, the PGW and the Node Exporter are legitimate uses of the injection hook. In both cases, you have `MetricFamily` protobuf objects already (for very different reasons). `Collector` collects `Metric` protobuf, so to emulate the injection hook in terms of a `Collector`, you needed to break the `MetricFamily` up, reverse-engineer `Desc` for them (which is where the registry gets the metric name and help from). Avoiding this tedious procedure was the reason (back then) for creating the injection hook in the first place. Thus, entirely removing the injection hook is not an option, even if the use case is rare.
- My argument above about "we need a custom registry for the Node Exporter" is actually invalid. Wisely, I have moved the logging concern into the handler, so all we need is a custom handler. Which makes the requirement to set an injection hook in the default registry a valid one again.
- One of my incentives for removing the setter was that it was never meant to be called after the first scrape. So the form was not in line with the semantics. But it is easy to bring the semantics in line with the form by allowing later setting of the injection hook (one mutex acquisition should do the trick).

So this is what I will do now:
- Reinstate ` SetMetricFamilyInjectionHook`, but in a concurrency-safe form, callable at any time. Big plus: One fewer breaking change.
- Make it a first-class citizen in the `Registry` interface and remove the special constructor. This makes the interface fatter, but the conclusions above show that it's essentially another type of registration, which is a core functionality of the registry.